packages/driver/src/cypress/ensure.ts

```diff
@@ -0,0 +1,314 @@
+
+// TODO: in 4.0 we should accept a new validation type called 'elements'
+// which accepts an array of elements (and they all have to be elements!!)
+// this would fix the TODO below, and also ensure that commands understand
+// they may need to work with both element arrays, or specific items
+// such as a single element, a single document, or single window
+
+// TODO: we should probably normalize all subjects
+// into an array and loop through each and verify
+// each element in the array is valid. as it stands
+// we only validate the first
+const validateType = (subject, type, name: string, cy: $Cy) => {
+  switch (type) {
+    case 'element':
+      // if this is an element then ensure its currently attached
+      // to its document context
+      if ($dom.isElement(subject)) {
+        isAttached(subject, name, cy)
+      }
+
+      // always ensure this is an element
+      return isElement(subject, name, cy)
+
+    case 'document':
+      return isDocument(subject, name, cy)
+
+    case 'window':
+      return isWindow(subject, name, cy)
+
+    default:
+      return
+  }
+}
+
+const isType = (subject, type, name: string, cy: $Cy) => {
+  let types: (string | boolean)[] = [].concat(type)
+
+  // if we have an optional subject and nothing's
+  // here then just return cuz we good to go
+  if (types.includes('optional') && _.isUndefined(subject)) {
+    return
+  }
+
+  // okay we either have a subject and either way
+  // slice out optional so we can verify against
+  // the various types
+  types = _.without(types, 'optional')
+
+  // if we have no types then bail
+  if (types.length === 0) {
+    return
+  }
+
+  let err
+  const errors: Error[] = []
+
+  for (type of types) {
+    try {
+      validateType(subject, type, name, cy)
+    } catch (error) {
+      err = error
+      errors.push(err)
+    }
+  }
+
+  function wordJoin (array) {
+    const copy = [...array]
+    const last = copy.pop()
+
+    return `${copy.join(', ')} or ${last}`
+  }
+
+  // every validation failed and we had more than one validation
+  if (errors.length === types.length) {
+    err = errors[0]
+
+    if (types.length > 1) {
+      // append a nice error message telling the user this
+      const msg = err.message.replace(/(failed because it requires .*)\./, `${wordJoin(['$1', ...types.slice(1)]) }.`)
+      const errProps = $errUtils.modifyErrMsg(err, '', () => msg)
+
+      $errUtils.mergeErrProps(err, errProps)
+    }
+
+    throw err
+  }
+}
+
+const isChildCommand = (command, args) => {
+  if (cy.subjectChain(command.get('chainerId')) === undefined) {
+    const stringifiedArg = $utils.stringifyActual(args[0])
+
+    $errUtils.throwErrByPath('miscellaneous.invoking_child_without_parent', {
+      args: {
+        cmd: command.get('name'),
+        args: _.isString(args[0]) ? `\"${stringifiedArg}\"` : stringifiedArg,
+      },
+    })
+  }
+}
+
+const isNotDisabled = (subject, name: string, onFail) => {
+  if (subject.prop('disabled')) {
```

Comment:
with exposing these publicly, any risk someone will pass a non-jquery element to `isNotDisabled`?

Reply:
Hm. Interesting thing to think about. The intended usage would be something like

```
Cypress.ensure.isElement(subject, 'myQuery')
Cypress.ensure.isNotDisabled(subject, 'myQuery')
```

If they don't ensure that it is an element first, and it's not an element, as it stands it would probably throw a pretty non-helpful error like "can't read property `prop` of null" or something.

If we added an implicit `isElement()` check to the top of each of these functions, that would provide more helpful error messages - but then we'd be running duplicate code for pretty much every cypress command.

I'm inclined to leave it as-is, but open to other thoughts.

Reply:
maybe we be a bit more guarded in our check? if the original conditions are not met, ignore it?
```
if (subject.prop && _.isFunction(subject.prop) && subject.prop('disabled') {
```

Reply:
though ignoring it sort of assumes it's true when it might be invalid? but oh well